[ADD]Chrome Compatibility

cguignol opened the pull request:
Mozilla API and Chrome API are different on several point + Behavior of extension in Chrome or FF is also different.

**Notice : I did not make tests of this code in Mozilla FF environment.**

All the modifications I made intend to support Chrome browser with few impact on existing code. Refactoring of this first attempt is required to have a "cross-browser" code compliant by its architecture... or by additional mecanisms (build, specific code ...).

- I made several modifications on **API calls** in script.js and popup.js

- I made strong modifications of **setByteLengthPerOrigin** to use a calculation way supported by Chrome (originUrl is missing from header read by Chrome +  browser.webRequest.filterResponseData is not supported by Chrome, due to security and architecture issues).

- I slightly changed popup.html behavior to support **"Shift project page" button**.

- I added several **Unit Tests** for script.js + pipeline configuration to run them directly in Github. They are exclusively dedicated for Chrome code, You can play them with npm on your computer.

Consider this code as a first try / "fast fail" to support Chrome ;)

Fix #1

cguignol committed:
[ADD]Chrome compatibility of the popup : link to shift project page working/use of Chrome API.

cguignol committed:
[ADD]Chrome compatibility of script.js : Chrome API called.
function headersReceivedListener uses initiator instead of originUrl to aggregate data in Domains.
headersReceivedListener counts the bytes of the page by using information contained directly in the ResponseHeader (filterResponseData is not supporter by Chrome).

cguignol committed:
[TESTS]Add Unit tests corresponding to script.js.

package.json contains all npm dependencies for testing (tests, mock, coverage).

cguignol committed:
Confirguration for testing

cguignol committed:
[FIX]Fixing test running

Remove useless build/ci instructions.

cguignol committed:
[FIX]test run

add npm install to get needed node_modules.

cguignol committed:
[UPD]Clean procedure

cguignol committed:
Merge pull request #1 from cguignol/cguignol-build-1

Github configuration for testing

cguignol commented:
Hi guys,

I made a very quick test this morning with my "Chrome" Code embedded into a temporary extension for Mozilla FF..

![image](https://user-images.githubusercontent.com/38350760/70130730-52904400-1681-11ea-96c2-e97c9e35fae1.png)

![image](https://user-images.githubusercontent.com/38350760/70131004-dcd8a800-1681-11ea-890f-7af0ef680d14.png)

**The result of FF seems to work.**

We have the "domain" problem (mozilla uses _OriginUrl_, Chrome uses _internal_), so **I'm going to make a quick fix to support this case**.

cguignol committed:
[FIX]FF compatibility of the chrome code evolution. FF uses "originUrl" to define the domain, Chrome uses "internal".

cguignol committed:
Merge branch 'master' of https://github.com/cguignol/Carbonalyser

supertanuki commented on script.js:
```diff
@@ -20,24 +20,18 @@ setByteLengthPerOrigin = (origin, byteLength) => {
+     const origin = extractHostname(!requestDetails.initiator ? (!requestDetails.originUrl ? requestDetails.url : requestDetails.originUrl) : requestDetails.initiator);
+     const rhf = requestDetails.responseHeaders.find(element => element.name.toLowerCase() === "content-length");
+     const rh = undefined === rhf ? {value: 0} 
+      : rhf;
```
Please use significant variable. requestDetails instead of rh, contentLengthHeader instead rhf ?

cguignol replied:
Done, thanks for the feedback ;)

cguignol committed:
[CLEANCODE]Explicit naming variables.

cguignol commented:
I made additional tests with site like https://www.750g.com/ : calculation is totally wrong with this kind of code and site.

First loading : 85 Mo.

Scrolling => 1,3 Go.

I'll watch this problem.

cguignol committed:
[UPD]Chrome extension : specific code is now isolated.
Mozilla FF uses its previous calculation mode based on filterResponseData, and Chrome uses content-length calculation.

supertanuki commented on script.js:
```diff
@@ -19,14 +19,33 @@ setByteLengthPerOrigin = (origin, byteLength) => {
   localStorage.setItem('stats', JSON.stringify(statsJson));
 }
 
+isChromeExtension = () => {
```
`isChrome` ?

cguignol replied:
OK, I am going to change it.

supertanuki commented on script.js:
```diff
@@ -19,25 +19,38 @@ setByteLengthPerOrigin = (origin, byteLength) => {
-    const origin = extractHostname(!requestDetails.originUrl ? requestDetails.url : requestDetails.originUrl);
-    setByteLengthPerOrigin(origin, event.data.byteLength);
+headersReceivedListener = (requestDetails) => {
+  if(isChromeExtension()){
```
Code styling
```suggestion
  if (isChromeExtension()) {
```

cguignol replied:
@supertanuki I can do this modification, but about formatting I have 2 options :

- make this specific modification.
- format the code with a code formatter like **Prettier** (included in my actual IDE VSCode), which lead to this specific modification + formatting the entire script.

What do you prefer?

supertanuki commented on script.js:
```diff
@@ -19,25 +19,38 @@ setByteLengthPerOrigin = (origin, byteLength) => {
+     const requestSize = new Number(contentLength.value); 
+     setByteLengthPerOrigin(origin, requestSize);
+  }
+  else {
```
```
    return {};
}
```
and remove the `else` ?

supertanuki commented on test/test-script.js:
```diff
@@ -0,0 +1,256 @@
+const testedScript = require('../script.js');
+
+
+
```
Delete empty lines ?

cguignol replied:
I have the same remark as here : https://github.com/supertanuki/Carbonalyser/pull/23#discussion_r361156686

cguignol committed:
[UPD]isChromeExtension() function renamed into isChrome()

supertanuki commented:
Thanks you @cguignol 
I'll prepare the release and publish it to the chrome store in the next days
